File: arrayAndStrings/checkPermutation/checkPermutation.py
'''
    Check Permutation: Given two strings, write a method to decide if one is a permutation of the
    other. 
    
    Time Complexity: O(n)
'''


def checkPermutation(stringA: str, stringB: str) -> bool:
    if(len(stringA) != len(stringB)):
        return False

    charHash = [0]*128  # Assuming String is Ascii (128 Characters)

    for char in stringA:
        charHash[ord(char)] += 1

    for char in stringB:
        charHash[ord(char)] -= 1
        if(charHash[ord(char)] < 0):
            return False

    return True


stringA = "Permutations"
StringB = "timutePnsrao"
print(checkPermutation(stringA, StringB))

File: arrayAndStrings/checkPermutation/checkPermutation.js
/*
    Check Permutation: Given two strings, write a method to decide if one is a permutation of the
    other. 
    
    Time Complexity: O(n)
*/


checkPermutations = (stringA, stringB) => {
    if (stringA.length !== stringB.length) {
        return false;
    }

    // Assuming String is Ascii (128 Characters)
    const charHash = new Array(128).fill(0); 

    for (let i = 0; i < stringA.length; i++) {
        const charCode = stringA.charCodeAt(i);
        charHash[charCode]++;
    }

    for (let i = 0; i < stringB.length; i++) {
        const charCode = stringB.charCodeAt(i);
        charHash[charCode]--;
        if (charHash[charCode] < 0) {
            return false;
        }
    }

    return true;
}


const stringA = "Permutations";
const StringB = "timutePnsrao";
console.log(checkPermutations(stringA, StringB));